.github/workflows/ci.yml

```diff
@@ -87,10 +100,26 @@ jobs:
       # Action for this, as of Jan 2021 it does not support Python coverage
       # files - only lcov files.  Therefore, we use coveralls-python, the
       # coveralls.io-supplied Python reporter, for this.
-      - name: "Report Coverage to Coveralls"
+      #
+      # It is coveralls-python 1.x that has maintained compatibility
+      # with Python 2, while coveralls-python 3.x is compatible with
+      # Python 3.  Sadly we can't use them both in the same workflow.
+      #
+      # The two versions of coveralls-python are somewhat mutually
+      # incompatible.  Mixing these two different versions when
+      # reporting coverage to coveralls.io will lead to grief, since
+      # they get job IDs in different fashion.  If we use them both in
+      # the same workflow, the finalizing step will be able to mark
+      # only part of the jobs as done, and the other part will be left
+      # hanging, never marked as done.  It doesn't matter if we make
+      # an API call or `coveralls --finish`.
+      #
+      # So we just use the newer coveralls-python that is available to
+      # Python 3 throughout this workflow.
+      - name: Report coverage to coveralls.io
         run: |
-          pip install coveralls
-          python -m coveralls
+          pip3 install --upgrade coveralls==3.0.1
+          python3 -m coveralls
```

Comment:
I guess what would be nice to see is the change in how coveralls is invoked split off into a separate ticket/PR from the change to add Python 3 coverage measurement/reporting.  Then it becomes more clear what the point of the different parts of this changeset are and it gets easier to understand the git history.

Reply:
That makes sense.  I will start a separate PR soon.